sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/Scan.java

```diff
@@ -102,4 +102,13 @@ default MicroBatchStream toMicroBatchStream(String checkpointLocation) {
   default ContinuousStream toContinuousStream(String checkpointLocation) {
     throw new UnsupportedOperationException(description() + ": Continuous scan are not supported");
   }
+
+  /**
+   * Returns an array of supported custom metrics with name and description.
+   * By default it returns empty array.
+   */
+  default CustomMetric[] supportedCustomMetrics() {
```

Comment:
maybe we should have a `SupportsReportMetrics` trait with this default implementation?

Reply:
Hmm, I'm okay on putting it into `SupportsReportMetrics`.

From implementation view, I don't see it is much different. Seems we just need to additionally check if a `Scan` is a `SupportsReportMetrics` when we want to get the supported metric list.

Semantically it sounds the same. A scan reports no custom metrics v.s. A scan does not support custom metrics.

Reply:
I also thought like @sunchao . +1 for a new trait.

Reply:
I can see the argument for a separate trait, but I would opt not to add this in one. We want to encourage sources to provide metrics. It's easier to see that there's a method that can be implemented than to see an interface that can be mixed in, especially when that interface might extend both `Scan` and `Write` (and not extend from either). There's also just more cost with having a lot of interfaces for very specific things.

I think an optional method is a good balance of concerns: it is optional, but easily discovered and doesn't add another interface that an implementer needs to know about and know which class to implement it in.

Reply:
it's a nit from me so I'm fine either way. About the method naming, should we make it shorter like `metrics`?

Reply:
Encourage is different from enforcing.
>  We want to encourage sources to provide metrics.

In addition, if we tightly couple this together, we are unable to split it permanently without a breaking change. That's a design risk to me.
> There's also just more cost with having a lot of interfaces for very specific things.

Given that, I'd recommend to put orthogonal concept separately.

Since this is a design issue, cc @cloud-fan , @gatorsmile , @gengliangwang , too.

Reply:
I'm fine for a separate trait. It is great if we can have a consensus so this can move forward. @rdblue Please let me know if you have strong option here.

Reply:
I'm okay with a separate trait. I wouldn't want to block this on a minor issue, but I don't think it is the right choice. If the general consensus is to go with a trait, please more forward.

Reply:
Thank you @rdblue.

Reply:
No problem! I didn't intend to delay this, I just thought that other people might agree with my reasoning.

Reply:
I don't see how a mixin trait helps avoid breaking change. Removing the trait later is also a breaking change. Agree with @rdblue that an optional method is simple and sufficient. Some other features have mixin trait because they don't have a reasonable default if made into optional methods. Metrics is OK as we can return empty metrics as the default.

Reply:
cc @dongjoon-hyun Does it convincing for you?

Reply:
I confirmed with @dongjoon-hyun offline and it's okay to move towards optional methods. Thanks all for the discussion.

Reply:
@cloud-fan Removing `trait` is a little different story. The point was we don't need to change `Scan` interface like this every time. For the use cases which only depends on `Scan` interface, we can guarantee that nothing is be broken.
> I don't see how a mixin trait helps avoid breaking change. Removing the trait later is also a breaking change.

However, I also don't want to block this PR like this for a long time. So, I switch to give +1 for @rdblue's suggestion.